### PR TITLE
feat: add Model Gateway usage analytics

### PR DESCRIPTION
**Pull request by Fermionic-Lyu**
## Summary

- add a dedicated Model Gateway Usage page with spend summaries, 30-day spend/request/token charts, and model/provider breakdowns
- simplify Overview to a single spend activity chart linking to the full Usage page
- add self-hosted Model Gateway Settings for independently updating the OpenRouter API key and management key
- keep cloud-hosted settings platform-managed while using the same dashboard usage contract

## Credential behavior

- Settings APIs return only masked credential status; the management key is never returned in plaintext and is used only for analytics
- the standard API key and management key are independent writes; partial success is retained, caches are cleared, and the dashboard refetches the resulting status
- self-hosting prefers encrypted stored credentials; `OPENROUTER_API_KEY` is seeded when possible and remains a fallback only when no stored API key exists
- cloud hosting continues to resolve managed credentials from InsForge Cloud

## Hosting coverage

- self-hosting calls OpenRouter with dashboard-configured encrypted credentials
- cloud hosting calls the InsForge Cloud backend; detailed model/provider activity depends on [insforge-cloud-backend#740](https://github.com/InsForge/insforge-cloud-backend/pull/740)
- deterministic hosting-boundary coverage is in [agent-e2e#10](https://github.com/InsForge/agent-e2e/pull/10)

## Validation

- backend: 152 files passed, 1818 tests passed, 3 files / 17 tests skipped
- dashboard: 38 files passed, 144 tests passed
- shared schemas build, monorepo typecheck, production builds, locale parity, OpenAPI parse, and changed-file formatting/lint passed
- repository CI, including lint/format, unit, component, integration, and E2E checks, passed
- release image: [v2.2.7-model-gateway-usage](https://github.com/InsForge/InsForge/actions/runs/29774032809)
- deterministic fixture E2E: [passed](https://github.com/InsForge/agent-e2e/actions/runs/29774193957)

---
## Summary by cubic
Adds a Model Gateway Usage page with 30‑day analytics and secure self‑hosted credential management, simplifies Overview to one spend chart, and hardens settings updates so partial successes persist and per‑field failure context is preserved even if audit logging fails.

- **New Features**
  - Usage page: 30‑day spend/requests/tokens charts and top models/providers with shares.
  - Overview: one spend chart linking to Usage; shared `AIActivityChartCard` with UTC dates, a11y, and unified currency.
  - Self‑hosted settings: masked status and independent updates for OpenRouter API and management keys; secrets encrypted; immediate cache invalidation; per‑field error details returned; successful key changes are preserved even if audit logging errors occur.
  - Admin APIs (self‑hosted): GET/PUT `/api/ai/config` with OpenAPI docs, schemas in `@insforge/shared-schemas`, and audit logging that records partial successes/failures.
  - Provider: `OpenRouterProvider` adds time‑series and per‑model usage (model/provider/BYOK), and `modelUsage` in overview.
  - Bootstrap: seed `OPENROUTER_API_KEY` once; stored secrets override env. New “Usage” sidebar and `/dashboard/ai/usage`; i18n (en/es/zh‑CN/zh‑TW).

- **Migration**
  - Self‑hosting: set `OPENROUTER_API_KEY` for first run; then manage keys from Model Gateway Settings. Stored secrets take precedence over env.
  - For detailed analytics, add the OpenRouter management key in settings (not used for inference).

<sup>Written for commit 9f7e4e3855d3bb997688e29312efcf81f08cbe52. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/InsForge/InsForge/pull/1761?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

> [!NOTE]
> ### Add Model Gateway usage analytics page and credential management for self-hosted deployments
> - Adds a new `/dashboard/ai/usage` page (`AIUsagePage`) showing spend summary, 30-day activity charts, and a top-10 model usage breakdown with per-model spend, tokens, and requests.
> - Introduces `ModelGatewayConfigService` to store and retrieve OpenRouter API and management keys in the encrypted secret store, seeding from environment variables on startup in non-cloud environments.
> - Adds `GET /api/ai/config` and `PUT /api/ai/config` endpoints (self-hosted only) so admins can read and update masked credentials; a `ModelGatewaySettingsDialog` in the sidebar exposes this in the UI.
> - Extends `OpenRouterProvider` to resolve the API key from the secret store and require a management key for activity/observability data; without a management key, activity charts show an unavailable state.
> - Adds `modelUsage` aggregation to the AI overview response, sorted by spend, using normalized activity items that now include model identity, provider, and BYOK spend fields.
> - Risk: self-hosted deployments without a management key configured will lose access to requests/tokens activity charts and model usage data until a key is set via Model Gateway Settings.
>
> 
> #### Changes since #1761 opened
>
> - Refactored Model Gateway credential management to remove source field distinction and implement environment variable fallback [6412adf]
> - Added audit logging for Model Gateway configuration updates [6412adf]
> - Fixed Authorization header construction for OpenRouter management activity fetch [6412adf]
> - Added localized reset cadence display and token breakdown tooltip [6412adf]
> - Added accessibility attributes to chart bars and enforced UTC timezone for all date formatting [6412adf]
> - Standardized currency formatting across usage charts [6412adf]
> - Added default observability error message and updated error text for missing API keys [6412adf]
> - Added OpenAPI specification for Model Gateway configuration endpoints [6412adf]
> - Updated environment variable documentation for OpenRouter API key [6412adf]
> - Added seed bootstrapping for OpenRouter API key from environment [6412adf]
> - Updated test suites to match new credential schema and behavior [6412adf]
> - Reformatted `ModelGatewayCredentialKey` type alias to multi-line style [75ac292]
> - Consolidated number and currency formatting utilities and replaced local implementations across AI dashboard components [c71146b]
> - Added error handling to `OpenRouterProvider.fetchActivity` method for management credential loading failures [c71146b]
> - Added non-whitespace validation pattern to API key fields in OpenAPI schema [c71146b]
> - Changed query invalidation timing in `useUpdateModelGatewayConfig` hook from `onSuccess` to `onSettled` [c71146b]
> - Added audit logging with success/failure tracking to the `/api/ai/config` PUT route handler in `modelGatewayConfigService` [072900c]
> - Added test coverage for audit logging behavior in AI configuration route tests [072900c]
> - Added clarifying comment to `ModelGatewayConfigService.getApiKey` method [f3f3474]
> - Introduced `ModelGatewayConfigUpdateError` class and refactored `ModelGatewayConfigService.updateConfig` method to track per-field success and failure outcomes [5839f44]
> - Modified PUT `/config` handler in ai.routes to distinguish partial success from complete failure in audit logging [5839f44]
> - Updated test coverage in `ai-routes.test.ts` and `model-gateway-config.service.test.ts` to verify partial success handling and per-field error reporting [5839f44]
> - Wrapped the `auditService.log` call in a try/catch block within the PUT `/api/ai/config` route handler to prevent audit logging failures from causing the request to error, logging the failure via `logger.error` instead and proceeding to return a successful response [0912d44]
> - Implemented partial success tracking and structured error propagation for Model Gateway configuration updates [9f7e4e3]
> - Added test coverage for partial success tracking and error cause preservation in Model Gateway configuration updates [9f7e4e3]
> 
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 8c66dae.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added an **AI Usage** page with spend summaries, spending-limit + reset cadence, activity charts, and top-model breakdowns.
  * Introduced **Model Gateway Settings** for self-hosted deployments, including an API key + management key dialog and a **Usage** navigation entry.
  * Added an admin **AI config** API endpoint to view/update masked OpenRouter credentials (with audit logging).
  * Updated AI overview to include **Model Usage** aggregation and a new spend/activity chart card.
* **Bug Fixes**
  * Improved self-hosted activity loading by enforcing management-key requirements and updating credential error guidance; added better handling for key changes.
* **Documentation**
  * Expanded `.env.example` to explain bootstrap/overwrite behavior for the OpenRouter credential.

### CODE REVIEW

**Review by jwfing — COMMENTED** [review]
## Review: Model Gateway usage analytics

**Summary:** A well-structured feature — new Usage page + analytics, encrypted secret-store credential management, and OpenRouter activity integration — with strong test coverage and sound credential handling; no blocking issues found.

### Requirements context
No spec/plan matched under `/docs/superpowers/` — `InsForge/InsForge` has no `docs/superpowers/` or `docs/specs/` directory (deploy/product docs live under `docs/`). Assessed against the PR description, the cubic/Macroscope summaries, and the surrounding code.

### Findings

#### Critical
(none)

#### Suggestion

**Functionality — self-hosted API key resolution has no request-time env fallback.** `backend/src/providers/ai/openrouter.provider.ts:150-162` now resolves the self-hosted key *exclusively* from the encrypted secret store; on `main` it read `process.env.OPENROUTER_API_KEY` directly. The env→store migration relies entirely on `seedApiKeyFromEnv()` at startup (`backend/src/utils/seed.ts:188-189`), and that method swallows errors (`backend/src/services/ai/model-gateway-config.service.ts:52-60,74-78`). If seeding fails (DB not ready at boot, `listSecrets`/`createSecret` throws), a previously-working `OPENROUTER_API_KEY` silently stops working with only a `warn` log, and the dashboard shows "not configured". The normal upgrade path is covered, so this is non-blocking — but consider a direct env fallback in `getApiKeyWithSource()` (or a louder diagnostic) for resilience.

**Performance/correctness — per-instance credential cache.** `model-gateway-config.service.ts:118-148` caches resolved credentials in-process for 60s, and `updateConfig` (line 114) only clears the *local* instance's cache. In a multi-instance self-hosted deployment, sibling instances serve stale credentials for up to 60s after a rotation. Low blast radius (self-hosted is usually single-instance), but worth noting.

#### Information

- **Dead credential-source branch.** `openrouter.provider.ts:161` maps `credential.source === 'environment' ? 'env' : 'settings'`, but `resolveStoredCredential` hardcodes `source: 'dashboard'` (`model-gateway-config.service.ts:150-154`), so `'environment'`/`'env'` is now unreachable for self-hosted. Functionally harmless (only affects error-hint wording), but the source no longer distinguishes env-seeded from dashboard-set keys.
- **Duplicate masking logic.** `maskCredential` (`model-gateway-config.service.ts:169-174`) and `maskApiKey` (`openrouter.provider.ts:198-203`) are identical — consider sharing one helper.
- **Raw timestamp in UI.** `KeyUsagePanel` renders the raw `limitReset` ISO string via `Resets {{value}}` (`packages/dashboard/src/features/ai/pages/AIUsagePage.tsx:112-118`); consider formatting it for display.
- **Activity hash heuristic.** `resolveActivityApiKeyHash` (`openrouter.provider.ts:426-429`) falls back to `sha256(apiKey)` when OpenRouter's `/key` `hash` is absent/malformed. This is an assumption about OpenRouter's hashing; if wrong it degrades gracefully to the "management key cannot access this key" state, so acceptable.

### Dimension notes
- **Software engineering:** Strong. New unit tests cover the config service (seed/update/mask, env-vs-stored precedence) and provider overview edge cases (no key, no management key, invalid key → business error, 30-day bucket boundaries, chronological sort, today/old-activity exclusion). Dashboard component tests added for the sidebar, settings dialog, and usage/overview pages. Conventions (Zod validation, `AppError`/`ERROR_CODES`, singleton services, `#`-alias imports) are followed.
- **Functionality:** Implements the described feature. `modelUsage` is optional in the schema (`packages/shared-schemas/src/ai-api.schema.ts:339`) for staged rollout, and the UI handles `observabilityAvailable === false` cleanly. See the Suggestion above re: the env-fallback removal.
- **Security:** Good. New `/api/ai/config` GET/PUT are `verifyAdmin` **and** gated to self-hosted (`assertSelfHostedModelGatewayConfig`, `index.routes.ts:151-159`); responses are masked-only and never echo plaintext; the update path never returns stored values. Credentials are stored as `isReserved` secrets, consistent with existing `JWT_SECRET`/Stripe handling. Cloud error bodies are logged but kept out of client-facing messages (`openrouter.provider.ts:589-625`). (The pre-existing `GET /:provider/api-key` endpoint returns the plaintext key for copy — unchanged by this PR, admin-gated.)
- **Performance:** No N+1 or hot-path regressions; the analytics fetch is on-demand and cached client-side (react-query, 60s) and server-side (60s). Only the multi-instance cache note above.

### Verdict
**approved** (informational) — zero Critical findings. Suggestions and Information items are non-blocking; the human approver still gives the explicit GitHub approval.

**Review by jwfing — DISMISSED** [review]
LGTM - approved.

File: backend/src/api/routes/ai/index.routes.ts [hunk 1]
```diff
@@ -8,17 +8,55 @@ import { AppError } from '@/utils/errors.js';
...
 const router = Router();
 const chatService = ChatCompletionService.getInstance();
+const modelGatewayConfigService = ModelGatewayConfigService.getInstance();
 type AIProvider = 'openrouter';
 
+/**
+ * GET /api/ai/config
+ * Get masked self-hosted Model Gateway credential status.
+ */
+router.get('/config', verifyAdmin, async (_req: AuthRequest, res: Response, next: NextFunction) => {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/ai/index.routes.ts:35]: Fixed in 6412adf: `openapi/ai.yaml` now documents GET/PUT `/api/ai/config`, admin security, validation/error responses, independent update semantics, and masked credential schemas.

File: backend/src/providers/ai/openrouter.provider.ts [hunk 2]
```diff
@@ -126,22 +147,25 @@ export class OpenRouterProvider {
...
       throw new AppError(
-        'OpenRouter API key not configured. Set OPENROUTER_API_KEY in the backend environment.',
+        'OpenRouter API key not configured. Set OPENROUTER_API_KEY or configure Model Gateway settings.',
         500,
         ERROR_CODES.AI_INVALID_API_KEY
       );
     }
-    return { apiKey, source: 'env' };
+    return {
+      apiKey: credential.value,
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/providers/ai/openrouter.provider.ts]: Not changing this behavior. The plaintext copy route predates this PR, is admin-protected, and is the product's intentional API-key copy capability. The new settings API itself returns only masked status, and no route exposes the management key in plaintext.

File: backend/src/services/ai/model-gateway-config.service.ts [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  async seedApiKeyFromEnv(): Promise<void> {
+    let existingKeys: Set<string>;
+    try {
+      const existingSecrets = await this.secretService.listSecrets();
+      existingKeys = new Set(existingSecrets.map((secret) => secret.key));
+    } catch (error) {
+      logger.warn('Failed to inspect existing Model Gateway credentials', {
+        error: error instanceof Error ? error.message : String(error),
+      });
+      return;
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/ai/model-gateway-config.service.ts:75]: Fixed in 6412adf: runtime API-key resolution now prefers the stored secret and falls back to `OPENROUTER_API_KEY` when the lookup succeeds with no stored key. Actual secret-store/decryption failures still propagate and are never cached as a missing key.

File: backend/src/api/routes/ai/index.routes.ts [hunk 4]
```diff
@@ -8,17 +8,55 @@ import { AppError } from '@/utils/errors.js';
...
+  } catch (error) {
+    next(error);
+  }
+});
+
+/**
+ * PUT /api/ai/config
+ * Store self-hosted Model Gateway credentials encrypted in system.secrets.
+ */
+router.put('/config', verifyAdmin, async (req: AuthRequest, res: Response, next: NextFunction) => {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/ai/index.routes.ts:48]: Fixed in 6412adf: route tests now cover successful GET/PUT behavior, invalid payloads, cloud GET/PUT rejection, audit metadata, and verify that secret values are never written to the audit log.

File: packages/dashboard/src/features/ai/pages/AIUsagePage.tsx [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    notation: 'compact',
+    maximumFractionDigits: 1,
+  }).format(value);
+}
+
+function formatChartCurrency(value: number): string {
+  return `$${value.toFixed(value >= 10 ? 0 : 2)}`;
+}
+
+function metricTotal(points: AIOverviewMetricPoint[]): number {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/ai/pages/AIUsagePage.tsx]: Keeping these helpers view-local. They are tiny pure format/reduce functions with page-specific presentation behavior; extracting them now would add a shared abstraction without reducing meaningful complexity. The concrete currency divergence was fixed by reusing the page's full-precision currency formatter.

File: backend/src/utils/seed.ts [hunk 6]
```diff
@@ -182,6 +183,12 @@ export async function seedBackend(): Promise<void> {
     // Initialize API key (from env or generate)
     const apiKey = await secretService.initializeApiKey();
 
+    // Self-hosted OpenRouter env values are bootstrap inputs. Seed them once so
+    // runtime reads and dashboard edits share the encrypted secret store.
+    if (!isCloudEnvironment()) {
+      await ModelGatewayConfigService.getInstance().seedApiKeyFromEnv();
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/utils/seed.ts:189]: Fixed in 6412adf: the `seedBackend()` test now exercises a missing stored secret plus `OPENROUTER_API_KEY` and asserts that the environment key is seeded through the startup path.

File: packages/dashboard/src/lib/i18n/locales/es.json [hunk 7]
```diff
@@ -508,9 +508,43 @@
...
+      "today": "Hoy",
+      "thisWeek": "Esta semana",
+      "thisMonth": "Este mes",
+      "allTime": "Todo el tiempo",
+      "spendingLimit": "Límite de gasto de la clave",
+      "spendingLimitUsed": "Límite de gasto usado de la clave de OpenRouter",
+      "of": "de",
+      "remaining": "Restante",
+      "limitReset": "Restablecimiento del límite",
+      "resetsValue": "Se restablece {{value}}",
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/lib/i18n/locales/es.json:527]: Fixed in 6412adf: known daily/weekly/monthly reset cadences are localized before interpolation, and ISO reset dates are formatted with the active locale in UTC.

File: backend/src/providers/ai/openrouter.provider.ts [hunk 8]
```diff
@@ -284,6 +301,16 @@ export class OpenRouterProvider {
       return this.fetchCloudActivity();
     }
 
+    const managementCredential = await this.modelGatewayConfigService.getManagementKey();
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/providers/ai/openrouter.provider.ts]: Not applicable by design. We intentionally removed the management-key environment variable in this PR: management credentials are dashboard-managed encrypted secrets only. `OPENROUTER_API_KEY` is the sole compatibility env fallback.

File: backend/src/providers/ai/openrouter.provider.ts [hunk 9]
```diff
@@ -284,6 +301,16 @@ export class OpenRouterProvider {
...
+    if (!managementCredential) {
+      return {
+        available: false,
+        data: [],
+        error:
+          'Configure an OpenRouter management key to load request, token, model, and historical activity.',
+      };
+    }
+
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/providers/ai/openrouter.provider.ts]: Fixed in c71146b: management-secret read/decryption failures are now caught inside the activity boundary, logged, and returned as an unavailable activity state. Key usage and the rest of Overview continue to render. Added provider coverage for this failure path.

File: packages/dashboard/src/features/ai/hooks/useModelGatewayConfig.ts [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      showToast(
+        error.message ||
+          t('ai.settings.saveFailed', { defaultValue: 'Failed to save Model Gateway settings' }),
+        'error'
+      );
+    },
+    onSettled: () => {
+      // A multi-key save may partially succeed because the two credentials are independent.
+      // Reconcile the masked status even when one write fails.
+      void queryClient.invalidateQueries({ queryKey: MODEL_GATEWAY_CONFIG_QUERY_KEY });
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/ai/hooks/useModelGatewayConfig.ts:45]: Fixed in c71146b: `onSettled` now invalidates config status, the standard OpenRouter key query, and AI overview for both success and partial failure. The success handler no longer duplicates those invalidations.

File: packages/dashboard/src/features/ai/pages/AIOverviewPage.tsx [hunk 11]
```diff
@@ -41,75 +38,43 @@ function formatModelCredit(value: number, compact = false): string {
   return `$${value.toFixed(2)}`;
 }
 
-function formatCompact(value: number): string {
-  return new Intl.NumberFormat('en', { notation: 'compact', maximumFractionDigits: 1 }).format(
-    value
-  );
+function formatChartCurrency(value: number): string {
+  return `$${value.toFixed(value >= 10 ? 0 : 2)}`;
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/ai/pages/AIOverviewPage.tsx]: Fixed in c71146b: Overview and Usage now share the same USD formatter, including cents for values above $10 and four decimals for non-zero sub-cent spend. The Overview test now covers `$12.34`.

File: packages/dashboard/src/features/ai/components/AIActivityChartCard.tsx [hunk 12]
```diff
@@ -0,0 +1,229 @@
+import { useTranslation } from 'react-i18next';
+import type { AIOverviewMetricPoint } from '@insforge/shared-schemas';
+
+function formatCompact(value: number): string {
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply packages/dashboard/src/features/ai/components/AIActivityChartCard.tsx]: Fixed in c71146b: compact-number, USD, and metric-sum helpers are now shared from the existing AI helpers module and reused by Overview, Usage, and the activity chart. The older unused token formatter was removed.

File: openapi/ai.yaml [hunk 13]
```diff
@@ -586,14 +634,91 @@ components:
...
+          nullable: true
+          description: Masked credential for display, or null when it is not configured.
+
+    ModelGatewayConfig:
+      type: object
+      required:
+        - apiKey
+        - managementKey
+      properties:
+        apiKey:
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply openapi/ai.yaml:657]: Fixed in c71146b: both OpenAPI credential fields now include a non-whitespace pattern matching the trimmed Zod contract.

File: backend/src/api/routes/ai/index.routes.ts [hunk 14]
```diff
@@ -8,17 +8,65 @@ import { AppError } from '@/utils/errors.js';
...
+    }
+    const config = await modelGatewayConfigService.updateConfig(validation.data);
+    await auditService.log({
+      actor: req.hasApiKey ? 'api-key' : req.user?.id,
+      action: 'UPDATE_MODEL_GATEWAY_CONFIG',
+      module: 'AI',
+      details: { updatedFields: Object.keys(validation.data) },
+      ip_address: req.ip,
+    });
+    successResponse(res, config);
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/ai/index.routes.ts]: Fixed in 072900c: validated configuration attempts are now audited on both success and failure with only field names plus an outcome. This preserves an audit trail for partial success without recording credential values; an audit failure in the error path is logged but does not mask the original credential error. Added route coverage for the failed/partial path.

**Comment by Fermionic-Lyu** [comment]
Addressed note 1 in f3f34740e. The new comment documents that OPENROUTER_API_KEY is only a fallback when the stored credential is absent; secret-store read/decryption failures intentionally propagate so they cannot silently revive a stale bootstrap key. This is documentation-only and does not change runtime behavior.

File: backend/src/api/routes/ai/index.routes.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/ai/index.routes.ts]: Follow-up in 5839f44be makes the partial-success audit exact. ModelGatewayConfigService now carries only succeeded/failed field names from Promise.allSettled; the route records updatedFields, failedFields, and outcome=partially_succeeded without credential values. Complete failures and successful updates use the same field-level audit shape. Added service and route coverage for both partial-success directions, both-write failure, and pre-write failure.

File: backend/src/api/routes/ai/index.routes.ts [hunk 15]
```diff
@@ -8,17 +8,84 @@ import { AppError } from '@/utils/errors.js';
...
+      throw error;
+    }
+    await auditService.log({
+      actor: req.hasApiKey ? 'api-key' : req.user?.id,
+      action: 'UPDATE_MODEL_GATEWAY_CONFIG',
+      module: 'AI',
+      details: { updatedFields, outcome: 'succeeded' },
+      ip_address: req.ip,
+    });
+    successResponse(res, config);
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/ai/index.routes.ts]: Fixed in 0912d4438. Success-path audit logging is now best-effort: an audit INSERT failure is logged server-side but no longer replaces an already completed credential update with a 500 response. Added route regression coverage proving updateConfig success plus audit rejection still returns 200 with the updated masked config.

File: backend/src/services/ai/model-gateway-config.service.ts [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+          return;
+        }
+
+        failedFields.push(field);
+        firstFailure ??= result.reason;
+      });
+
+      if (failedFields.length > 0) {
+        const message = firstFailure instanceof Error ? firstFailure.message : String(firstFailure);
+        throw new ModelGatewayConfigUpdateError(
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/services/ai/model-gateway-config.service.ts:155]: Fixed in 9f7e4e385. After recording the per-field settlement, the route now rethrows ModelGatewayConfigUpdateError.cause rather than the wrapper, preserving the original AppError identity, HTTP status, error code, and nextActions without duplicating metadata. Tests cover a partial success with a 409 SECRET_ALREADY_EXISTS cause while audit logging also fails, plus the real service wrapper retaining that AppError cause.

File: backend/src/api/routes/ai/index.routes.ts [hunk 17]
```diff
@@ -8,17 +8,95 @@ import { AppError } from '@/utils/errors.js';
...
+    }
+    const updatedFields = Object.keys(validation.data);
+    let config: Awaited<ReturnType<typeof modelGatewayConfigService.updateConfig>>;
+    try {
+      config = await modelGatewayConfigService.updateConfig(validation.data);
+    } catch (error) {
+      const succeededFields =
+        error instanceof ModelGatewayConfigUpdateError ? error.succeededFields : [];
+      const failedFields =
+        error instanceof ModelGatewayConfigUpdateError ? error.failedFields : updatedFields;
```
[the remark this answers is not in the document]
Comment (Fermionic-Lyu) [reply backend/src/api/routes/ai/index.routes.ts:67]: Fixed comprehensively in 9f7e4e385. The earlier automatic Addressed marker on 0912d44 was not sufficient: updateConfig now wraps post-write read-back failures with succeededFields containing every committed field and failedFields empty. The route therefore audits the mutation outcome as succeeded while still returning the underlying read-back error because the response config could not be constructed. Added service and route regression coverage.

**Review by jwfing — APPROVED** [review]
LGTM - approved.
